keras/callbacks.py

```diff
@@ -2791,10 +2790,13 @@ def on_train_batch_end(self, batch, logs=None):
                 step=self._train_step,
             )
 
-        # `logs` isn't necessarily always a dict. For example, when using
-        # `tf.distribute.experimental.ParameterServerStrategy`, a
-        # `tf.distribute.experimental.coordinator.RemoteValue` will be passed.
-        # For now, we just disable `update_freq` in those cases.
+        should_record = tf.equal(self._train_step % self.update_freq, 0)
```

Comment:
Unfortunately doing this with ParameterServerStrategy will result in unnecessary syncs across the cluster within an epoch. Using a thread to periodically log the metrics manually would work better for PSS. Or, alternatively, we've planned on time-based (asynchronous) callbacks but they're not available yet.

Thanks for the PR!

Reply:
? I’m measured the performance of this solution with actual large training jobs and if update frequency is kept high (100) I don’t see any speed impact at all. I also checked in debugger and while logs is a remote value, update step and update freq are not remote values and do not appear to trigger a sync.

Reply:
<img width="1792" alt="Screenshot 2023-01-19 at 9 33 45 AM" src="https://user-images.githubusercontent.com/16809055/213518274-98700108-ee53-41bd-82a5-51ce49c76406.png">
Succeeded (1 hr 3 min)

vs

<img width="1792" alt="Screenshot 2023-01-19 at 9 35 28 AM" src="https://user-images.githubusercontent.com/16809055/213518622-dd89fa54-a8a7-4d57-a1f6-5a7ba9086671.png">
Succeeded (59 min)

The first is training with Parameter Server strategy without this patch. The second is with this patch. The performance impact is less then noise in training speed.

Reply:
Understood - I think it's expected that with a high number of `update_freq`, we do see minimal performance impact because it's far enough between syncs, but they do sync when `logs.get()` is called.

An issue here is that with the asynchronous training with `ParameterServerStrategy`, the time at which the `on_batch_end` is called is not really when a batch is finished executing, but when a batch is finished being scheduled to be executed on a worker. That is, if you have `update_freq=500`, the callback is called when the 500th batch has been just "scheduled", at which point very few train functions may have been executed if the function is taking a long time, so it could be a misleading metric. And, I felt having a time-based treatment of this may be less confusing. What do you think?

Reply:
I think time based async callback is better, but it's also question of when is that likely to come? This solution has weaknesses, but is small change and works reasonably if you tune update_freq. There are other callbacks usable in parameter server with similar need to tune update_freq like ModelCheckpoint callback. Ups and downs in cpu utilization from the job I showed mainly come from model checkpointing.

I can update documentation to be clearer that with PS you should pick high update_freq for performance.

Right now without this pr, tensorboard callback feels unusable on Parameter server with long epochs. I commonly train models for only 1 or 2 epochs, so epoch level metrics give me almost no visualization of model metrics.

Reply:
Thanks for the comment and your patience! Let me bring this to the team and get back to you with an eta on time-based callback. We're worried calling the batch-level calls between scheduling can be misleading to users, so for now I think the best solution would be for you to have a subclassed TensorBoard callback for your specific use case. Thanks for the understanding!

Reply:
I'll go ahead and close this given response. If there is desire to re-visit this/add batch metric support to tensorboard callback happy to re-open then (or have news about time based callbacks).